sigmatch: simpler procedural type matching

zerbina opened the pull request:
## Summary

* simplify the logic for handling generic routines using `auto` as the
  return type
* reject un-instantiated routines earlier when used in branch and array
  constructor expressions, resulting in clearer error messages
  (`has no concrete type` instead of `cannot instantiate`)
* fix spurious `cannot instantiate` errors when passing generic routines
  as arguments to parameters that are also generic
* support lambda expressions coming from template expansions

## Details

Introduced by https://github.com/nim-lang/Nim/pull/3234, when both the
formal and actual type are unresolved meta types, the relation is
considered to be `isBothMetaConvertible`. The intention seems to have
been supporting passing `auto`-returning routines or lambda-expressions
to generic procedure-type parameters, but the logic introduced by the
aforementioned PR is not necessary for achieving that.

`auto` return types now use dedicated handling in `procTypeRel`, both
the formal and actual type being unresolved generics results in a
mismatch again, and `typeRel` is no longer applied repeatedly for
`isBothMetaConvertible` matches.

There was also the issue of `generateTypeInstance` being misused, which
could result in spurious `cannot instantiate` or internal compiler
errors. Until type instantiation properly supports the case where it's
not certain whether all used generic parameters have been inferred
already, the `tryGenerateInstance` routine is introduced as a temporary
solution.

Finally, both `nkStmtListExpr` and `nkBlockExpr` nodes are now
considered when instantiating a generic routine as part of parameter
matching, which allows for complex expressions yielding un-instantiated
generic routines (e.g.: `call((discard x; genericRoutine))`).

zerbina committed:
sigmatch: simpler procedural type matching

Summary
=======

* simplify the logic for handling generic routines using `auto` as the
  return types
* reject un-instantiated routines earlier when used  in branch and array
  constructor expressions, resulting in clearer error messages
  (`has no concrete type` instead of `cannot instantiate`)
* support lambda expressions coming from template expansions

Details
=======

Introduced by https://github.com/nim-lang/Nim/pull/3234, when both the
formal and actual type are unresolved meta types, the relation is
considered to be `isBothMetaConvertible`. The intention seems to have
been supporting passing `auto`-returning routines or lambda-expressions
to generic procedure-type parameters, but the logic introduced by the
aforementioned PR is not necessary for achieving that.

`auto` return types now use dedicated handling in `procTypeRel`, both
the formal and actual type being unresolved generics results in a
mismatch again, and `typeRel` is no longer applied repeatedly for
`isBothMetaConvertible` matches.

There was also the issue of `generateTypeInstance` being misused, which
could result in spurious `cannot instantiate` or internal compiler
errors. Since meta types are possibly involved, the correct routine to
use is `prepareMetatypeForSigmatch`. Testing that the produced
type instance is not a meta type was also done incorrectly (`isMetaType`
was used instead of `containsGenericType`).

Finally, `nkStmtListExpr` nodes are now considered when instantiating a
generic routine as part of parameter matching, which allows for complex
expressions yielding un-instantiated generic routines (e.g.:
`call((discard x; genericRoutine))`).

zerbina commented:
Regarding the effect on overload quality, the only difference is for generic procedural types with an `auto` return type. If after instantiation, the concrete procedural type matches exactly, the match is now considered as 1 conversion + 1 generic instead of 1 conversion + 1 exact. Example case:
```nim
proc call(x: proc(x: int): int {.nimcall.}) = discard

proc generic[T](x: T): auto {.nimcall.} =
  result = x

call(generic) # after instantiation, the types match exactly
```

I'm unsure whether this is a good idea or not.

saem commented on compiler/sem/sem.nim:
```diff
@@ -243,6 +243,23 @@ proc fitNodeConsiderViewType(c: PContext, formal: PType, arg: PNode; info: TLine
   else:
    result = a
 
+proc genericProcCheck(c: PContext, n: PNode): PNode =
```
It only generiates an error if the expression is a generic routine.

Would this be a better name: `exprNotGenericRoutine`?

saem commented on compiler/ast/ast_types.nim:
```diff
@@ -1768,10 +1768,9 @@ type
-                             ## the proc's return value
+    isBothMetaConvertible    ## a generic procedure that otherwise matched has
+                             ## an 'auto' return type; it needs to be instantiated
+                             ## first
```
```suggestion
    isBothMetaConvertible    ## a generic procedure with an 'auto' return type
                             ## that otherwise matched; it needs to be
                             ## instantiated first
```

I thought separating the exceptional aspect (auto return type) from the description of the subject of the match (generic routine with auto return) makes it easier to follow.

saem commented on compiler/sem/semtypinst.nim:
```diff
@@ -793,6 +793,10 @@ proc recomputeFieldPositions*(t: PType; obj: PNode; currPosition: var int) =
                            t: PType): PType =
+  ## Produces the instantiated type for the generic type `t`, using the
+  ## bindings provided by `pt`. All type variables used by `t` must have
+  ## *concrete* type bounds to them -- both meta types and missing bindings
```
```suggestion
  ## *concrete* types bound to them -- both meta types and missing bindings
```

saem commented on compiler/sem/sigmatch.nim:
```diff
@@ -509,20 +510,20 @@ proc procParamTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
+  ## two different calls:
+  ## - matches `f`=T to `a`=Q. `f` is a resolved metatype ('int' is bound to
+  ##   it already), so `a` can be inferred; a binding for Q=int is saved
+  ## - matces `f`=U to `a`=Q. `f` is an unresolved metatype, but since Q was
```
```suggestion
  ## - matches `f`=U to `a`=Q. `f` is an unresolved metatype, but since Q was
```

saem commented on compiler/sem/sigmatch.nim:
```diff
@@ -539,12 +541,14 @@ proc procParamTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
-        return isBothMetaConvertible
+      if f.isNil() or containsGenericType(f):
+        # no luck resolving the type, so the inference fails. Note that using
+        # ``isMetaType`` wont work, as it doesn't check for nested
```
```suggestion
        # ``isMetaType`` won't work, as it doesn't check for nested
```

saem commented on compiler/sem/sigmatch.nim:
```diff
@@ -2078,6 +2086,62 @@ template matchesVoidProc(t: PType): bool =
+    # possible in a concept context. There's nothing to instantiate
+    return nil
+  else:
+    # nothing else is able to provide uninstantiated generic routines
```
(just thinking outloud; entirely ignorable)

This isn't a bug, just an observation, sigmatch has an assumption that all name lookups have been complete and we're dealing with symbols only (not counting untyped AST parameter values, but sigmatch just passes those around like opaque values).  This is one of those invariants of the module that I'm figuring out how to capture in the long run (entirely outside this PR).

saem commented on compiler/sem/sem.nim:
```diff
@@ -243,6 +243,23 @@ proc fitNodeConsiderViewType(c: PContext, formal: PType, arg: PNode; info: TLine
+
+  # skip all statement list wrappers:
+  var it {.cursor.} = n
+  while it.kind == nkStmtListExpr:
```
Do we also need to consider block expressions?

zerbina replied:
Hmm, good question. From the top of my head, I can't think of a reason why we don't want to support that case.

zerbina replied:
They're now also considered. Thanks for raising this inconsistency!

saem commented on tests/lang_callable/overload/tproc_type_inference.nim:
```diff
@@ -0,0 +1,95 @@
+  proc generic[T](x: T): auto =
+    result = x
+
+  call(1, (discard ""; generic))
+  # also works with lambda expressions:
+  call(2, (discard ""; (proc(x: auto): auto = x)))
```
```suggestion
  call(1, (discard ""; generic))
  call(2):
    block:
      discard ""
      generic
  # also works with lambda expressions:
  call(3, (discard ""; (proc(x: auto): auto = x)))
  call(4):
    block:
      discard ""
      (proc(x: auto): auto = x)
```

Just to illustrate the block expression case.

saem commented on tests/lang_callable/overload/tproc_type_inference.nim:
```diff
@@ -0,0 +1,95 @@
+
+  call(1, generic)
+  # a lambda expression with two parameters using the same type variable
+  # (i.e., generic parameter) doesn't work
```
(no action needed) I'm still tracing through figuring out why this is the case, but I'm guessing this is due to a pre-existing issue that begins with the `semGenerateInstance` call within `instantiateRoutineExpr`? If you don't know that's ok, just want to make sure I'm "digging in the right spot".

zerbina replied:
It's because a lambda expression with explicit generic parameters is not supported by the parser.

```nim
call(1, (proc[T](x: T, y: T) = discard))
```
is not valid code.

saem replied:
Thanks -- it's always a bit surprising what isn't supported. Although it might be tricky in ways I don't yet realize.

saem reviewed (approved):
This is far more sensible than before. 🎉 

I made some doc suggestions and possibly discovered a case that needs covering (`nkBlockExpr`).

saem commented:
> Regarding the effect on overload quality, the only difference is for generic procedural types with an `auto` return type. If after instantiation, the concrete procedural type matches exactly, the match is now considered as 1 conversion + 1 generic instead of 1 conversion + 1 exact. Example case:
> 
> ```nim
> proc call(x: proc(x: int): int {.nimcall.}) = discard
> 
> proc generic[T](x: T): auto {.nimcall.} =
>   result = x
> 
> call(generic) # after instantiation, the types match exactly
> ```
> 
> I'm unsure whether this is a good idea or not.

It's not ideal, but I'm not quite sure if +1 generic and exact is either. If it's cumbersome to fix right now I would note it in a comment as an issue and move on.

zerbina commented:
I have reverted the effects an instantiated `auto`-returning generic has on an overloads quality back to how they were before this PR.

The change with `prepareMetatypeForSigmatch` was also not quite correct:
- it doesn't always produce proper instances (because the instance cache is partially disabled), which causes subtle issues and can lead to C compiler errors
- the `isMetaType` + `containsGenericType` check isn't able to detect generic phantom types like, for example, `Type[T] = object`, which would lead to compiler assertions (rightfully) failing

To more properly fix the misuse of `generateTypeInstance` for the time being, I introduced the `tryGenerateInstance` routine. It uses `prepareMetatypeForSigmatch` as a way to check if all referenced generic parameters have been inferred already, and only if that is the case creates a proper instance via `generateTypeInstance`. This is rather inefficient, but it's the most robust and low-impact temporary solution that I could come up with.

zerbina committed:
address review comments

* fix typos
* rename `genericProcCheck` to `exprNotGenericRoutine`

zerbina committed:
sigmatch: support block expressions

Block expressions that produce a generic routine are now also
supported.

zerbina committed:
sigmatch: restore the previous counter behaviour

A procedural type that matches exactly after the procedure was
instantiated now counts towards exact matches again, instead of towards
generic ones.

zerbina committed:
tests: add test for the failure case

zerbina committed:
tests: move the tests to the `typerel` category

No overload resolution is taking place in the tests.

In addition, slightly adjust the "return type of formal procedural type
inference" case in a way so that it better catches inference failures.

zerbina committed:
sigmatch: properly resolve the formal type

`generateTypeInstance` is not intended to be used when it's not certain
that all generic parameters have a concrete type bound -- it's misuse
led to spurious instantiation failure errors.

It first has to be verified that all used generic parameters resolve to
concrete types, which is for now achieved via first calling
`prepareMetatypeForSigmatch` and then checking that the produced type
is not a generic type.

zerbina commented:
That should fix the test failures. I had forgotten that the `tfHasMeta` flag is also not *excluded* properly, meaning that it cannot be relied on at all for types produced by `replaceTypeVarsT` when meta-types are allowed.

saem commented:
/merge

saem commented:
This is a good step forward in making this bit of sem make more sense.
